[VDG] Spectrum control optimisation

wieslawsoltes opened the pull request.

molnard commented:
See my observations below:

- The animation seems to be a static picture. 
- I received an exception after 4 minutes

> System.AccessViolationException
>   HResult=0x80004003
>   Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
>   Source=<Cannot evaluate the exception source>
>   StackTrace:
> <Cannot evaluate the exception stack trace>
> 

![image](https://user-images.githubusercontent.com/9844978/171827496-823da89c-2c5f-4aff-91e5-f9776e3ffd7a.png)

### Idea

Since the background of the animation is a rectangle with a solid color basically the whole animation can be cached in ~50 static frames and displayed after each other. There is no need to "generate" the animation every time - just once all the frames at the beginning and then just display in a loop.

wieslawsoltes commented:
> See my observations below:
> 
> * The animation seems to be a static picture.

That's because of reduced update internal of frame rendered, looks like it's not workable with input data we have.

> * I received an exception after 4 minutes
> 
> > System.AccessViolationException
> > HResult=0x80004003
> > Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
> > Source=
> > StackTrace:
> 
> ![image](https://user-images.githubusercontent.com/9844978/171827496-823da89c-2c5f-4aff-91e5-f9776e3ffd7a.png)
> 

Looks like I can not cache SKPaint, need to create every frame, need to investigate.

> ### Idea
> Since the background of the animation is a rectangle with a solid color basically the whole animation can be cached in ~50 static frames and displayed after each other. There is no need to "generate" the animation every time - just once all the frames at the beginning and then just display in a loop.

The main issue is to know when source data is invalidated, right now we just pull data constantly and render it, if we can get some source data invalidation mechanism I could probably try caching frames.

soosr commented:
I have problems with this solution.
- The window is being redrawn ~15FPS (on my machine) all the time
- Start CJ (wait for animation) then stop it. The GPU usage stays above 1%.
- The animation is way less active... due to the FPS decrease and If I modify this code:
```
DispatcherTimer.Run(
	() =>
	{
		InvalidateVisual();
		return true;
	},
	TimeSpan.FromMilliseconds(50),
	DispatcherPriority.Render);
```
`TimeSpan.FromMilliseconds(50)` to `TimeSpan.FromMilliseconds((float)1000/60)` meaning invalidate it on 60 FPS instead of 20 FPS. So this results in the splash animation is not laggy anymore, and the activity of the CJ animation is better. 

Without CJ animation:
- Window is being redrawn on 40 FPS
- GPU usage is 1.6%

With CJ animation:
- Window is being redrawn on 40 FPS
- GPU usage is 3%

Although I haven't checked to code but I don't see improvement. GPU usage is lower because it renders on lower FPS, and it is even worse because the GPU is used all the time.

wieslawsoltes commented:
> I have problems with this solution.
> 
> * The window is being redrawn ~15FPS (on my machine) all the time
> * Start CJ (wait for animation) then stop it. The GPU usage stays above 1%.
> * The animation is way less active... due to the FPS decrease and If I modify this code:
> 
> ```
> DispatcherTimer.Run(
> 	() =>
> 	{
> 		InvalidateVisual();
> 		return true;
> 	},
> 	TimeSpan.FromMilliseconds(50),
> 	DispatcherPriority.Render);
> ```
> 
> `TimeSpan.FromMilliseconds(50)` to `TimeSpan.FromMilliseconds((float)1000/60)` meaning invalidate it on 60 FPS instead of 20 FPS. So this results in the splash animation is not laggy anymore, and the activity of the CJ animation is better.
> 
> Without CJ animation:
> 
> * Window is being redrawn on 40 FPS
> * GPU usage is 1.6%
> 
> With CJ animation:
> 
> * Window is being redrawn on 40 FPS
> * GPU usage is 3%
> 
> Although I haven't checked to code but I don't see improvement. GPU usage is lower because it renders on lower FPS, and it is even worse because the GPU is used all the time.

Changed FPS to 60 and added visibility check.

soosr commented:
Significant improvement! Great job! During CJ:
![image](https://user-images.githubusercontent.com/16364053/171981829-64fe94e7-a263-4403-885b-ebf647ebf60d.png)

@yahiheb Could you test this PR?

soosr commented on WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs:
```diff
@@ -55,6 +55,18 @@ public SpectrumControl()
 			}
 		};
+
+		DispatcherTimer.Run(
+			() =>
+			{
+				if (IsVisible)
+				{
+					InvalidateVisual();
+				}
+				return true;
+			},
+			TimeSpan.FromMilliseconds((float)1000/60),
+			DispatcherPriority.Render);
```
Would it make sense to not run it all the time, but when the control is visible?

```
renderTimer = new DispatcherTimer(DispatcherPriority.Render) { Interval = TimeSpan.FromMilliseconds((float)1000 / 60) };
renderTimer.Tick += (_, _) => InvalidateVisual();
```

And call start when IsVisible changed to true and call stop when changed to false?

soosr commented on WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs:
```diff
@@ -140,28 +152,61 @@ public override void Render(DrawingContext context)
 		context.Custom(this);
 
-		Dispatcher.UIThread.Post(() => InvalidateVisual());
+		// Dispatcher.UIThread.Post(() => InvalidateVisual());
```
can be removed if not needed anymore

wieslawsoltes replied:
removed

soosr commented on WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs:
```diff
@@ -55,6 +55,18 @@ public SpectrumControl()
 			}
 		};
+
+		DispatcherTimer.Run(
+			() =>
+			{
+				if (IsVisible)
+				{
+					InvalidateVisual();
+				}
+				return true;
+			},
+			TimeSpan.FromMilliseconds((float)1000/60),
+			DispatcherPriority.Render);
```
Would it make sense to not run it all the time, but when the control is visible?

```
renderTimer = new DispatcherTimer(DispatcherPriority.Render) { Interval = TimeSpan.FromMilliseconds((float)1000 / 60) };
renderTimer.Tick += (_, _) => InvalidateVisual();
```

And call start when IsVisible changed to true and call stop when changed to false?

wieslawsoltes replied:
You trade running timer for another observable, not sure it will save us much.

yahiheb commented:
> Significant improvement! Great job! During CJ: ![image](https://user-images.githubusercontent.com/16364053/171981829-64fe94e7-a263-4403-885b-ebf647ebf60d.png)
> 
> @yahiheb Could you test this PR?

On my machine I don't see much difference.


Master:

![Cajhjhj](https://user-images.githubusercontent.com/52379387/171989026-d803f43d-a174-4d38-82ea-6784a6fea2e4.PNG)

PR:

![pppp](https://user-images.githubusercontent.com/52379387/171989027-bf5a8d1a-52ec-4288-a939-7f4b8b9cbae3.PNG)

yahiheb commented:
@kiminuo Can you also test this?

soosr commented:
@yahiheb when you run the software in your logs between the first few messages you can find the Renderer line, can you share yours here?

wieslawsoltes commented:
> > Significant improvement! Great job! During CJ: ![image](https://user-images.githubusercontent.com/16364053/171981829-64fe94e7-a263-4403-885b-ebf647ebf60d.png)
> > @yahiheb Could you test this PR?
> 
> On my machine I don't see much difference.
> 
> Master:
> 
> ![Cajhjhj](https://user-images.githubusercontent.com/52379387/171989026-d803f43d-a174-4d38-82ea-6784a6fea2e4.PNG)
> 
> PR:
> 
> ![pppp](https://user-images.githubusercontent.com/52379387/171989027-bf5a8d1a-52ec-4288-a939-7f4b8b9cbae3.PNG)

What is your GPU ?

yahiheb commented:
> @yahiheb when you run the software in your logs between the first few messages you can find the Renderer line, can you share yours here?

@soosr  `ANGLE (Intel(R) UHD Graphics Direct3D11 vs_5_0 ps_5_0)`

> What is your GPU ?

@wieslawsoltes https://github.com/zkSNACKs/WalletWasabi/issues/7455#issuecomment-1078529552

wieslawsoltes commented:
> I've been playing with this and I think there are several things that would improve CPU/GPU use:
> 
> 1. We keep creating new and new objects to dispose them immediately - like [here](https://github.com/wieslawsoltes/WalletWasabi/blob/45544f548d10e62d7be9bfd09e68b941c751b911/WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs#L193-L197):
>    
>    * that filter (`var filter = SKImageFilter.CreateBlur(24, 24, SKShaderTileMode.Clamp);`) does not depend on anything variable so it can be created once. It's a `class` so we allocate memory 60 times per second, seems wasteful.
>    * `SKPictureRecorder` looks like it can be re-used too.
>    * The idea is [kiminuo@27cb38f](https://github.com/kiminuo/WalletWasabi/commit/27cb38f80af094e6e8b87868ac581041b75292b8). Though, I'm not sure when `void IDisposable.Dispose()` is being called and it breaks the code. Caching of skia should be possible according to the [comment](https://github.com/mono/SkiaSharp/issues/829#issuecomment-504835148).
>    * @wieslawsoltes Do you know please when that `Dispose()` method is called? I can see in the call stack that it's called from external code only.
> 2. Cache the animation itself. WIP commit [b076983](https://github.com/zkSNACKs/WalletWasabi/commit/b07698311094bb52a8c70d63ecffd3c02dc6aac3). It's not done but it just shows the idea.
>    
>    * One might say that it's impossible because `AuraSpectrumDataSource` uses random numbers. I think it would be worth a try to generate say 100 random vectors and cache the output for those. Then unless user changes the app window dimensions, there is no need to compute anything and one just redraws already computed canvases (or something). It might be slightly less random but honestly having not-really-random animation vs. my computer will keep burning my knees.. I pick the former :)

If we want cache we should cache only recorded SKPicture as it's the sole purpose, caching other skia primitives is not good idea as you can get issues like David got with memory access violations. Anything that goes into recording picture should not be reused and shall be disposed.

wieslawsoltes commented:
If we want to cache frames we want to do it here using resulting picture:
https://github.com/zkSNACKs/WalletWasabi/blob/45544f548d10e62d7be9bfd09e68b941c751b911/WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs#L226

yahiheb commented on WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs:
```diff
@@ -55,6 +58,46 @@ public SpectrumControl()
+							frameCounter = 0;
+						}
+					}
+					else
+					{
+						if (frameCounter >= EffectLength)
+						{
+							isEffectActive = false;
+							frameCounter = 0;
+							InvalidateVisual();
+						}
+					}
+
+					if (isEffectActive || IsFireEffectVisible)
+					{
+						InvalidateVisual();
+					}
```
Small optimization:

```suggestion
					else
					{
						if (frameCounter >= EffectLength)
						{
							isEffectActive = false;
							frameCounter = 0;
						}
						
						InvalidateVisual();
					}

					if (IsFireEffectVisible)
					{
						InvalidateVisual();
					}
```

wieslawsoltes replied:
Added, thanks.

wieslawsoltes commented on WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs:
```diff
@@ -55,6 +58,46 @@ public SpectrumControl()
 		};
+
+		const int EffectRepeatInterval = 15 * 60;
+		const int EffectLength = 5 * 60;
```
Moved and added comment

wieslawsoltes commented on WalletWasabi.Fluent/Controls/Spectrum/SpectrumControl.cs:
```diff
@@ -21,8 +21,8 @@ public class SpectrumControl : TemplatedControl, ICustomDrawOperation
 
-	private ImmutablePen? _linePen;
 	private IBrush? _lineBrush;
+	private SKColor _lineColor;
```
Removed

soosr commented:
I have like 10% of GPU usage spikes. For me, the constant ~0.6% looked better.

soosr commented:
> > I have like 10% of GPU usage spikes. For me, the constant ~0.6% looked better.
> 
> Another alternative is to run the animation first few seconds and then freeze it. Then we won't waste CPU & GPU. We can improve it in WW 2.1 to find good balance when we are not in time pressure.

I think exactly this is happening now.

> I have like 10% of GPU usage spikes. For me, the constant ~0.6% looked better.

It seems if I use my laptop screen then the GPU usage is around 0.6% if my external screen is connected, then it goes up to 10% when the animation starts to move. I am sure it is not because of this PR, but Avalonia. Is this normal? @wieslawsoltes

yahiheb commented:
GPU disabled:

![GPU disabled 0](https://user-images.githubusercontent.com/52379387/172962220-79f5064c-30b4-4989-83b9-fdb230f5d0b3.PNG)

GPU enabled:

![GPU enabled 0](https://user-images.githubusercontent.com/52379387/172962232-a958ad9e-117c-40d5-a7f9-8ddee1cfffc3.PNG)

wieslawsoltes commented:
> GPU disabled:
> 
> ![GPU disabled 0](https://user-images.githubusercontent.com/52379387/172962220-79f5064c-30b4-4989-83b9-fdb230f5d0b3.PNG)
> 
> GPU enabled:
> 
> ![GPU enabled 0](https://user-images.githubusercontent.com/52379387/172962232-a958ad9e-117c-40d5-a7f9-8ddee1cfffc3.PNG)

Looks like the Intel integrated GPU is really slow (it might be driver issue, Intel GPU are known to have serious problems with drivers and know to be really bad at performance), even CPU can do rendering much faster. Can you try running on dedicated GPU

soosr commented:
I have discussed https://github.com/zkSNACKs/WalletWasabi/pull/8234/commits/f63dc3e65608d9ab778e16669696165de1cd052d with @molnard, it is ok to be here.

@wieslawsoltes restart message should work now.

wieslawsoltes commented:
> I have discussed [f63dc3e](https://github.com/zkSNACKs/WalletWasabi/commit/f63dc3e65608d9ab778e16669696165de1cd052d) with @molnard, it is ok to be here.

Great.

> 
> @wieslawsoltes restart message should work now.

Tested, works great!

molnard commented:
I think this PR is a mess, merging two features together. Can we clean up that? GPU switch is a standalone feature, isn't it?

wieslawsoltes commented:
> I think this PR is a mess, merging two features together. Can we clean up that? GPU switch is a standalone feature, isn't it? Or is that related to the optimization?

It's related to this PR, after we finish testing I can split the GPU switch into separate PR

molnard commented:
As far as I can see EnableGPU is just to pass an argument at the beginning to Avalonia.

wieslawsoltes commented:
@molnard Opened PR with only Enable GPU setting included https://github.com/zkSNACKs/WalletWasabi/pull/8342

yahiheb commented:
> Looks like the Intel integrated GPU is really slow (it might be driver issue, Intel GPU are known to have serious problems with drivers and know to be really bad at performance), even CPU can do rendering much faster. Can you try running on dedicated GPU

@wieslawsoltes These numbers are taken when the spikes happen.

**Running on dedicated GPU**

GPU disabled:

![Disabled GPU 1](https://user-images.githubusercontent.com/52379387/173206817-998ee525-5b29-4c72-8cd5-9f13f3d49d8d.PNG)

GPU enabled:

![Enabled GPU 2](https://user-images.githubusercontent.com/52379387/173206820-a6281ba9-723c-4690-8f7a-eda9c8158769.PNG)

nopara73 commented:
I mistakenly posted my review of this PR here: https://github.com/zkSNACKs/WalletWasabi/pull/8342#pullrequestreview-1003651301

MaxHillebrand reviewed (dismissed):
It feels smoother, but I didn't run proper benchmarks.

molnard commented:
What's up with this, can be merged?

wieslawsoltes commented:
> What's up with this, can be merged?

This is ready. Dan had merged further optimization into master already that should give big improvements, so it would be good to re-check master GPU/CPU usage and if't still giving issues we can merge this too.

molnard commented:
> > What's up with this, can be merged?
> 
> This is ready. Dan had merged further optimization into master already that should give big improvements, so it would be good to re-check master GPU/CPU usage and if't still giving issues we can merge this too.

I think only @zkSNACKs/visual-design-group can make a relevant benchmark and decide on this.

wieslawsoltes commented:
> > > What's up with this, can be merged?
> > 
> > 
> > This is ready. Dan had merged further optimization into master already that should give big improvements, so it would be good to re-check master GPU/CPU usage and if't still giving issues we can merge this too.
> 
> I think only @zkSNACKs/visual-design-group can make a relevant benchmark and decide on this.

We can wait with merging this PR

yahiheb reviewed (commented):
> > I have like 10% of GPU usage spikes. For me, the constant ~0.6% looked better.
> 
> Another alternative is to run the animation first few seconds and then freeze it. Then we won't waste CPU & GPU. We can improve it in WW 2.1 to find good balance when we are not in time pressure.

I guess @kiminuo suggestion makes the most sense here, we can run the animation once for a while then freeze it for good.

This would make everyone happy, and of course after the release it can be further improved.

soosr commented:
What's up with this? Is it waiting for something or waiting for a final review?

wieslawsoltes commented:
> What's up with this? Is it waiting for something or waiting for a final review?

This was backup solution before releasing 2.0, I don't think this is best solution. We should explore another options after Avalonia 11 new renderer is ready.

wieslawsoltes commented:
Closing this as it's not the best solution, only to be used when we really need to reduce GPU/CPU usage, otherwise as I stated above we should wait for Avalonia 11 new renderer.
